library/modules/EventManager.cpp

```diff
@@ -193,1087 +240,1564 @@ std::array<eventManager_t,EventType::EVENT_MAX> compileManagerArray() {
-            newTickQueue.insert(pair<int32_t,EventHandler>(tick+(*i).first, (*i).second));
-        tickQueue.clear();
-        tickQueue.insert(newTickQueue.begin(), newTickQueue.end());
-        //out.print("%s,%d: on load, frame_counter = %d\n", __FILE__, __LINE__, tick);
-        */
-        //tickQueue.clear();
-        if (!df::global::item_next_id)
-            return;
-        if (!df::global::building_next_id)
-            return;
-        if (!df::global::job_next_id)
-            return;
-        if (!df::global::ui)
-            return;
+static event_tracker<InteractionData> interactionEvents; // we track ids as POD, and so we cannot be certain of the life cycle of what those ids reference
+
+static InteractionData getAttacker(color_ostream& out, df::report* attackEvent, df::unit* lastAttacker,
+                                   df::report* defendEvent, const vector<df::unit*>& relevantUnits);
+static std::vector<df::unit*> gatherRelevantUnits(color_ostream& out, df::report* r1, df::report* r2);
+
+/** Move Map
+ * scan: sendInvasionEvents
+ *
+ * scan_units: sendUnitNewActiveEvents
+ * scan_units: sendUnitDeathEvents
+ * scan_units: sendSyndromeEvents
+ * scan_units: sendInventoryChangeEvents
+ * scan_unit_inventory: sendInventoryChangeEvents
+ * scan_unit_reports: updateReportToRelevantUnits
+ *
+ * scan_items: sendItemCreationEvents
+ *
+ * scan_jobs: sendJobInitiatedEvents
+ * scan_jobs: sendJobStartedEvents
+ * scan_jobs: sendJobCompletedEvents
+ *
+ * scan_buildings: sendBuildingEvents
+ * scan_buildings: sendConstructionEvents
+ *
+ * scan_reports: sendReportEvents
+ * scan_reports: sendInteractionEvents
+ * scan_reports: sendUnitAttackEvents
+ * parse_strike_report: sendUnitAttackEvents
+ * */
+
+class Scanner{
+private:
+    Scanner() = default;
+    // current tick's data
+    std::unordered_set<int32_t> all_units;
+    std::unordered_set<int32_t> active_units;
+    std::unordered_set<int32_t> valid_items;
+    std::unordered_set<int32_t> valid_reports;
+    std::unordered_set<int32_t> valid_buildings;
+    std::unordered_map<int32_t, df::job*> valid_jobs;
+    std::unordered_map<df::coord, df::construction*> valid_constructions;
+
+public:
+    const std::unordered_map<int32_t, df::job*> &current_jobs = valid_jobs;
+    inline void scan(color_ostream& out, const int32_t &tick) {
         if (!df::global::world)
             return;
+        scan_units(out, tick);
+        scan_items(out, tick);
+        scan_jobs(out, tick);
+        scan_buildings(out, tick);
+        scan_reports(out, tick); // depends on scanning units first
 
-        nextItem = *df::global::item_next_id;
-        nextBuilding = *df::global::building_next_id;
+        // if there's a new invasion we'll add it
+        if (!df::global::ui)
+            return;
+        invasions.emplace(tick, nextInvasion);
         nextInvasion = df::global::ui->invasions.next_id;
-        lastJobId = -1 + *df::global::job_next_id;
+    }
+    inline void clear() {
```

Comment:
EventManager is basically intended to centralize scanning data structures for changes, and centralizing inside the implementation is just a good next step to that. So this can be turned into a separate PR. Though this will need rewriting as this is intended to also prepare sending on multiple tick frequencies.